hazelcast/src/test/java/com/hazelcast/jet/core/IndeterminateSnapshotTest.java

```diff
@@ -0,0 +1,1285 @@
+                logger.info(thisAddress + " sending sync packet (type=" + type + ") to " + endpoint + " action: " + action);
+            } else {
+                action = Action.ALLOW;
+            }
+
+            return action;
+        }
+    }
+
+    @Ignore("test processor used by EE tests")
+    public static final class SnapshotInstrumentationP extends AbstractProcessor {
+        static final ConcurrentMap<Integer, Integer> savedCounters = new ConcurrentHashMap<>();
+        static final ConcurrentMap<Integer, Integer> restoredCounters = new ConcurrentHashMap<>();
+        static final Set<Long> executions = ConcurrentHashMap.newKeySet();
+
+        private int globalIndex;
+        private final int numItems;
+        /**
+         * Number of completed snapshots.
+         */
+        private int snapshotCounter;
+        /**
+         * Id of first snapshot for which {@link #saveSnapshotConsumer} and other hooks are called.
+         * Snapshot ids are assigned starting with 0.
+         */
+        public static volatile int allowedSnapshotsCount = 0;
+        @Nullable
+        public static volatile Consumer<Integer> saveSnapshotConsumer;
+        @Nullable
+        public static volatile Consumer<Integer> snapshotCommitPrepareConsumer;
+        @Nonnull
+        public static volatile Consumer<Boolean> snapshotCommitFinishConsumer = b -> { };
+        public static volatile boolean restoredFromSnapshot = false;
+
+        SnapshotInstrumentationP(int numItems) {
+            this.numItems = numItems;
+        }
+
+        /**
+         * Reset global processor settings and state to default values
+         */
+        public static void reset() {
+            savedCounters.clear();
+            restoredCounters.clear();
+            executions.clear();
+
+            allowedSnapshotsCount = 0;
+            saveSnapshotConsumer = null;
+            snapshotCommitPrepareConsumer = null;
+            snapshotCommitFinishConsumer = b -> { };
+            restoredFromSnapshot = false;
+        }
+
+        @Override
+        protected void init(@Nonnull Context context) {
+            globalIndex = context.globalProcessorIndex();
+            executions.add(context.executionId());
+        }
+
+        @Override
+        public boolean complete() {
+            // emit current snapshot counter
+            // finish when requested number of items was emitted
+            return tryEmit(10000 * (globalIndex + 1) + snapshotCounter)
+                    && snapshotCounter == numItems;
+        }
+
+        @Override
+        public boolean saveToSnapshot() {
+            if (saveSnapshotConsumer != null && snapshotCounter >= allowedSnapshotsCount) {
+                saveSnapshotConsumer.accept(globalIndex);
+            }
+            savedCounters.put(globalIndex, snapshotCounter);
+            return tryEmitToSnapshot(broadcastKey(globalIndex), snapshotCounter);
```

Comment:
Since we're emitting exactly one item to the snapshot queue, there will always be room for it, but this code will handle the case when the queue is full better:
```suggestion
            if (!tryEmitToSnapshot(broadcastKey(globalIndex), snapshotCounter)) {
                return false;
            }
            if (saveSnapshotConsumer != null && snapshotCounter >= allowedSnapshotsCount) {
                saveSnapshotConsumer.accept(globalIndex);
            }
            savedCounters.put(globalIndex, snapshotCounter);
            return true;
```